consensus/consensus_service.go

```diff
@@ -77,24 +78,30 @@ func (consensus *Consensus) UpdatePublicKeys(pubKeys, allowlist []bls_cosi.Publi
 	// TODO: use mutex for updating public keys pointer. No need to lock on all these logic.
 	consensus.pubKeyLock.Lock()
 	consensus.Decider.UpdateParticipants(pubKeys, allowlist)
-	consensus.getLogger().Info().Msg("My Committee updated")
-	for i := range pubKeys {
-		consensus.getLogger().Info().
-			Int("index", i).
-			Str("BLSPubKey", pubKeys[i].Bytes.Hex()).
-			Msg("Member")
-	}
-
 	allKeys := consensus.Decider.Participants()
+	consensus.pubKeyLock.Unlock()
 	if len(allKeys) != 0 {
-		consensus.LeaderPubKey = &allKeys[0]
+		first := consensus.Decider.FirstParticipant(
+			shard.Schedule.InstanceForEpoch(consensus.Blockchain.CurrentHeader().Epoch()))
+		consensus.pubKeyLock.Lock()
+		consensus.LeaderPubKey = first
+		consensus.pubKeyLock.Unlock()
 		consensus.getLogger().Info().
 			Str("info", consensus.LeaderPubKey.Bytes.Hex()).Msg("My Leader")
 	} else {
 		consensus.getLogger().Error().
 			Msg("[UpdatePublicKeys] Participants is empty")
 	}
-	consensus.pubKeyLock.Unlock()
+	for i := range pubKeys {
+		consensus.getLogger().Info().
+			Int("index", i).
+			Str("BLSPubKey", pubKeys[i].Bytes.Hex()).
+			Msg("Member")
+	}
+	if consensus.Blockchain.Config().IsLeaderRotation(consensus.GetCurEpoch()) {
+		consensus.updateLeader()
```

Comment:
when the committee (publicKeys) are updated, that means the epoch is switching, we should just default back to the first key as the leader. Why call updateLeader() here?